문나혜: 섹션8,9의 미션 제출

moonnahye opened the pull request:
## #️⃣연관된 이슈

#62

## 📝학습 내용

### 정적 중첩클래스
- static이 붙는다
- 바깥 클래스의 인스턴스에 소속되지 않는다.

**중첩클래스는 언제 사용하나?**
- 특정 클래스가 다른 하나의 클래스 안에서만 사용되는 경우
- 캡슐화: 중첩클래스는 바깥 클래스의 private 멤버에 접근할 수 있다. 이렇게 해서 둘을 긴밀하게 연결하고 불필요한 public 메서드를 제거할 수 있다. 

### 내부 클래스
- static이 붙지않는다
- 바깥 클래스의 인스턴스에 소속된다.
- 바깥 클래스의 인스턴스, 클래스 멤버에 접근할 수 있다.
  - 지역 클래스 
    - 지역 클래스는 내지역 변수와 같이 코드 블럭 안에서 정의된다.
  - 익명 클래스
    - 이름이 없는 지역 클래스이다.
    - 특정 부모 클래스(인터페이스)를 상속 받고 바로 생성하는 경우 사용한다.
    - 지역 클래스가 일회성으로 사용되는 경우나 간단한 구현을 제공할 때 사용한다. 


### 스크린샷 (선택)

## 💬리뷰 요구사항

지역클래스 익명클래스를 구현하려고 하니 생각보다 이해한걸 나타내는게 쉽지않아서, 제대로 한건지 모르겠네요.
수정하면 좋을것같다는 부분은 알려주시면감사하겠습니당!

moonnahye committed:
feat. 섹션 8,9 미션 제출

moonnahye committed:
feat. 섹션 8,9 미션 수정

SungbinYang commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,164 @@
+                return table;
+            }
+        }
+        return null;
```
현재 getAvailableTable 메서드는 예약 가능한 테이블이 없을 경우 null을 반환하고 있습니다. 이는 Optional을 사용하는 방식으로 변경하면 null 처리와 관련된 문제를 줄일 수 있습니다.

moonnahye replied:
수정해보겠습니다!

SungbinYang commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,164 @@
+    private Map<Integer, Table> tables = new HashMap<>();
+    private Map<Integer, Reservation> reserveList = new HashMap<>();
+
+    static class Table {
```
Table 클래스에 toString 메서드가 없어 디버깅 및 출력 시 정보 확인이 어렵습니다.

moonnahye replied:
넵! 테이블 전체조회 메서드도 만들어봐야겠습니당

SungbinYang commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,164 @@
+        return null;
+    }
+
+    void displayAvailableTableList() {
```
displayAvailableTableList 메서드는 예약 가능한 테이블이 없는 경우와 있는 경우를 별도로 처리하고 있습니다. 이를 좀 더 간결히 작성할 수 있지 않을까요?

moonnahye replied:
수정해보겠습니다!

SungbinYang reviewed (commented):
간단하게 리뷰 남겨드렸습니다! 확인바랍니다!

sm9171 commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,164 @@
+
+public class Restaurant {
+
+    private Map<Integer, Table> tables = new HashMap<>();
+    private Map<Integer, Reservation> reserveList = new HashMap<>();
```
테이블과 예약을 map형식으로 담아서 중복을 제거해주었군요!

sm9171 reviewed (approved):
잘 구현해주셨네요.
수고하셨습니다~!

moonnahye committed:
fix. 코드 리뷰 반영 및 기능 추가 by 성빈님

- getAvailableTable() 메서드 반환값 옵셔널 처리
- 테이블 toString() 추가 및 displayAllTableList() 메서드 추가
- findReservationsByDate() 메서드 추가
- findReservationsByName() 메서드 추가

tdronnie commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    public void createTable(int tableNumber) {
+        if (tables.containsKey(tableNumber)) {
+            throw new IllegalArgumentException("이미 존재하는 테이블 번호입니다.");
+        }
```
테이블 번호는 대부분 1번에서 시작해서 차례대로 증가할 것 같은데
사용자에게 랜덤의 테이블 번호를 받는 자유를 줘도 되는지에 대한 생각을 해보게 되는 것 같습니다!

tdronnie commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,192 @@
+
+
+    public void reserve(int reservationNumber, String reserveDate, String reserveName, String comment) {
+        class TableStatusUpdater {
+            void markAsReserved(Table table, int reservationNumber) {
+                table.tableState = TableState.RESERVED;
+                table.reserveNumber = reservationNumber;
+            }
+        }
```
아래의 cancel메서드도 해당하는 것인데 지역 클래스에 대한 로직을 하나의 메서드로 묶어서 statusUpdate에 관련한 코드를 한곳에서 관리해도 좋을 것 같습니다!

tdronnie replied:
제가 생각해보았는데 결국 지역 클래스를 사용하려면 각 메서드에서 정의해야 할 것 같아요
공통 로직으로 분리하고 클래스 인스턴스를 반환받을 수 있을 것 같았는데 안되더라구요
혹시 좋은 방법 있으시면 공유 부탁드립니다!

tdronnie reviewed (approved):
코멘트 남겨드렸습니다
지역클래스와 익명클래스에 대해서는 잘 이해되도록 작성해주신 것 같아요!

hmkim199 commented on nahye/java-mid1/src/section08/TableState.java:
```diff
@@ -0,0 +1,5 @@
+package section08;
+
+public enum TableState {
```
열거형으로 구현해주신 거 좋네요! OCP원칙에 맞게 나중에 예약 취소 상태나 대기가 추가되더라도 수정에 용이할 것 같아요. 👍

hmkim199 commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,192 @@
+            this.comment = comment;
+        }
+
+        private void parseDate(String reserveDate) {
```
지난 시간에 배웠던 날짜와 시간 라이브러리도 잘사용해주신 것 같아요!

hmkim199 reviewed (approved):
구체적으로 잘 구현해주셨네요!! 예약이 시간 단위로도 가능하면 더 좋을 것 같아요! 고생하셨습니다 :)

SungbinYang reviewed (approved):
수정하신 부분 확인했습니다!

Darimi2222 commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,192 @@
+public class Restaurant {
+
+    private Map<Integer, Table> tables = new HashMap<>();
+    private Map<Integer, Reservation> reserveList = new HashMap<>();
```
테이블과 예약을 모두 Map으로 관리하게 설계하셨군요! 👍

Darimi2222 commented on nahye/java-mid1/src/section08/Restaurant.java:
```diff
@@ -0,0 +1,192 @@
+    static class Table {
+        private int tableNumber;
+        private TableState tableState;
+        private Integer reserveNumber;
```
예약 번호를 래퍼 클래스로 선언한 이유가 있을까요?

Darimi2222 reviewed (commented):
간단한 코멘트 남기고 갑니다!
